src/macaron/slsa_analyzer/checks/provenance_available_check.py

```diff
@@ -78,6 +74,281 @@ def __init__(self) -> None:
         ]
         super().__init__(check_id=check_id, description=description, depends_on=depends_on, eval_reqs=eval_reqs)
 
+    def find_provenance_assets_on_package_registries(
+        self,
+        repo_full_name: str,
+        package_registry_data_entries: list[PackageRegistryData],
+        provenance_extensions: list[str],
+    ) -> Sequence[Asset]:
+        """Find provenance assets on package registries.
+
+        Note that we stop going through package registries once we encounter a package
+        registry that does host provenance assets.
+
+        Parameters
+        ----------
+        repo_full_name : str
+            The full name of the repo, in the format of ``owner/repo_name``.
+        package_registry_data_entries : list[PackageRegistryData]
+            A list of package registry data entries.
+        provenance_extensions : list[str]
+            A list of provenance extensions. Assets with these extensions are assumed
+            to be provenances.
+
+        Returns
+        -------
+        Sequence[Asset]
```

Comment:
There are a couple of reasons:
* The indirect reason is I want to return objects of type `Asset` instead of `JFrogMavenAsset` from this method. This return type of `Sequence[Asset]` is also consistent with the method `find_provenance_assets_on_ci_services`. The `run_check` method - the caller of these `find_provenance_assets_*` methods - does not need to know of the `JFrogMavenAsset` type or whatever concrete asset type they return. The `Asset` interface is enough.
* The direct reason for `Sequence[Asset]` instead of `list[Asset]` is Python's type variance rule for generic container types. What this method really returns is a `list[JFrogMavenAsset]`. `JFrogMavenAsset` is a subtype of `Asset`. However, because `list` (like other mutable containers in Python) is invariant, a `list[JFrogMavenAsset]` is not necessarily a `list[Asset]`. Meanwhile, `Sequence` is covariant, and therefore a `Sequence[JFrogMavenAsset]` is also a `Sequence[Asset]`.

[This page in the mypy doc](https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types) gives a good example of why a multiple containers like `list` should be invariant. Here is the example they use:
```python
class Circle(Shape):
    # The rotate method is only defined on Circle, not on Shape
    def rotate(self): ...

def add_one(things: list[Shape]) -> None:
    things.append(Shape())

my_circles: list[Circle] = []
add_one(my_circles)     # This may appear safe, but...
my_circles[-1].rotate()  # ...this will fail, since my_circles[0] is now a Shape, not a Circle
```
In this example, if the function `add_one` is annotated as `add_one(things: Sequence[Shape])`, then the method call `things.append` would be prohibited by the type checker, preventing the issue in the first place. `Sequence` is a supertype of both `list` and `tuple`, therefore it cannot be mutated because it will violate the fact that the container can be a `tuple`. While mutable containers like `list` are invariant, immutable containers like `tuple` are covariant, and as a result `Sequence` is also covariant.

Sorry for the wall of text, but hopefully the explanation clarifies my intention. 😄